### PR TITLE
Instrument the front-end

### PR DESCRIPTION
**Pull request by bp323**
The main conceptual approach here is to instrument all human initiated interactions in the app, organised around the main critical flows so that we have quantifiable data with contextual information on task completion rates and times, so that event funnels can be created the product can be iteratively improved based on these metrics.

All the stuff mentioned below assumes a 3rd party data warehouse and visualisation service in place like [Mixpanel](https://mixpanel.com) We could write our own tracking code or could also look at leveraging [segment.io](https://segment.com/) to abstract the instrumentation JS code and turn on integration to Mixpanel AND Google Analytics with 2 clicks.

**Documentation**
The list below contains the list of instrumentation points. Every single entry should be part of the documentation on Github and notes should be put in for each individual event on WHEN the events is sent + other remarks. 
Example: Cancel event -> "Sent when either the Cancel link or the modal close dialog icon is clicked or the user clicked outside the modal window"
This might look cumbersome, but it will pay off on the long run when people will interpret the data 

**Generic points**
In all events we send the following 
 * user_type: ["Superadmin", "Departmental administrator", "Lecturer", "Student"]
 * logged_in: ["True", "False"]
 * user: hash("on215", "key", "salt") for "Student" & "Lecturer" user types
 * create Mixpanel user profiles for "Superadmin" and "Departmental administrator" user types and track them

**Student UI instrumentation points**

 * [x] SUI Generic - API call error [API endpoint, http_code]

 * [x] SUI Auth - Signed in
 * [x] SUI Auth - Signed in error [http_code]
 * [x] SUI Auth - Sign out

 * [x] SUI Navigation - Tripos selector - Opened
 * [x] SUI Navigation - Tripos selector - Search [final search term]
 * [x] SUI Navigation - Tripos selector - Selected [triposID]
 * [x] SUI Navigation - Part selector - Opened
 * [x] SUI Navigation - Part selector - Selected [triposID, partID, time_from_start]
 * [x] SUI Navigation - Module opened
 * [x] SUI Navigation - Module closed
 * [x] SUI Navigation - Draft timetable page shown
 * [x] SUI Navigation - Cambridge Logo clicked
 * [x] SUI Navigation - Authentication modal triggered
 * [x] SUI Navigation - Knowledge base clicked
 * [x] SUI Navigation - Terms & Conditions clicked

 * [x] SUI Calendar - View - Week view selected
 * [x] SUI Calendar - View - Week - Week <
 * [x] SUI Calendar - View - Week - Week >
 * [x] SUI Calendar - View - Week - Term >
 * [x] SUI Calendar - View - Week - Term <
 * [x] SUI Calendar - View - Week - Today
 * [x] SUI Calendar - View - Month view selected
 * [x] SUI Calendar - View - Month - Month <
 * [x] SUI Calendar - View - Month - Month >
 * [x] SUI Calendar - View - Month - Term >
 * [x] SUI Calendar - View - Month - Term <
 * [x] SUI Calendar - View - Month - Today
 * [x] SUI Calendar - View - Day view selected
 * [x] SUI Calendar - View - Day - Day <
 * [x] SUI Calendar - View - Day - Day >
 * [x] SUI Calendar - View - Day - Term >
 * [x] SUI Calendar - View - Day - Term <
 * [x] SUI Calendar - View - Day - Today
 * [x] SUI Calendar - View - Tooltip displayed
 * [x] SUI Calendar - Manage - Module added [time_from_start]
 * [x] SUI Calendar - Manage - Module removed
 * [x] SUI Calendar - Manage - Series added [time_from_start]
 * [x] SUI Calendar - Manage - Series removed
 * [x] SUI Calendar - Print clicked
 * [x] SUI Calendar - Export - Opened
 * [x] SUI Calendar - Export - Closed
 * [x] SUI Calendar - Export - Subscribed to calendar feed [ics_feed_hash, time_from_start]
 * [x] SUI Calendar - Export - Other ways - Opened
 * [x] SUI Calendar - Export - Other ways - Closed
 * [x] SUI Calendar - Export - Other ways - URL Clicked [time_from_start]
 * [x] SUI Calendar - Export - Other ways - Google Calendar help clicked [time_from_start]
 * [x] SUI Calendar - Export - Other ways - MS Outlook help  clicked [time_from_start]
 * [x] SUI Calendar - Export - Other ways - Apple Calendar help clicked [time_from_start]

**Admin UI**
 * [x] AUI Generic - API call error [API endpoint, http_code]

 * [x] AUI Auth - Signed in
 * [x] AUI Auth - Signed in error [http_code]
 * [x] AUI Auth - Sign out

 * [x] AUI Navigation - Home - Shortcut clicked
 * [x] AUI Navigation - Home - Intro video played
 * [x] AUI Navigation - Home - Intro video collapsed
 * [x] AUI Navigation - Tripos selector - Opened
 * [x] AUI Navigation - Tripos selector - Search [final search term]
 * [x] AUI Navigation - Tripos selector - Selected [triposID]
 * [x] AUI Navigation - Part selector - Opened
 * [x] AUI Navigation - Part selector - Selected [triposID, partID]
 * [x] AUI Navigation - Home icon clicked
 * [x] AUI Navigation - Module opened [moduleID]
 * [x] AUI Navigation - Module closed [moduleID]
 * [x] AUI Navigation - Series opened [seriesID, is_borrowed, can_be_edited]
 * [x] AUI Navigation - Calendar view selected [partID]
 * [x] AUI Navigation - List view selected [partID]

 * [x] AUI Manage - Add module - Started
 * [x] AUI Manage - Add module - Cancelled
 * [x] AUI Manage - Add module - Completed [time_from_started]
 * [x] AUI Manage - Add series - Started
 * [x] AUI Manage - Add series - Cancelled
 * [x] AUI Manage - Add series - Completed [time_from_started]
 * [x] AUI Manage - Borrow series - Started
 * [x] AUI Manage - Borrow series - Chosen tripos
 * [x] AUI Manage - Borrow series - Chosen part
 * [x] AUI Manage - Borrow series - Module opened
 * [x] AUI Manage - Borrow series - Module closed
 * [x] AUI Manage - Borrow series - Series added
 * [x] AUI Manage - Borrow series - Series removed
 * [x] AUI Manage - Borrow series - Cancelled
 * [x] AUI Manage - Borrow series - Completed [time_from_start, number_of_borrowed_series]
 * [x] AUI Manage - Rename series via menu
 * [x] AUI Manage - Rename series inline
 * [x] AUI Manage - Delete series [is_borrowed, number_of_modules_using_this]
 * [x] AUI Manage - Remove borrowed series
 * [x] AUI Manage - Publishing - Started
 * [x] AUI Manage - Publishing - Cancelled
 * [x] AUI Manage - Publishing - Completed [time_from_start, date_stamp]

 * [x] AUI Data - Added event
 * [x] AUI Data - Added events for empty term
 * [x] AUI Data - Term selected [is_empty]
 * [x] AUI Data - Event selected
 * [x] AUI Data - All events selected from menu
 * [x] AUI Data - Title edit - Started [is_empty]
 * [x] AUI Data - Title edit - Completed
 * [x] AUI Data - DateTime edit - Started
 * [x] AUI Data - DateTime edit - Calendar item selected
 * [x] AUI Data - DateTime edit - Week changed
 * [x] AUI Data - DateTime edit - Day changed
 * [x] AUI Data - DateTime edit - Start hour changed
 * [x] AUI Data - DateTime edit - Start minute changed
 * [x] AUI Data - DateTime edit - End hour changed
 * [x] AUI Data - DateTime edit - End minute changed 
 * [x] AUI Data - DateTime edit - Completed [time_from_start]
 * [x] AUI Data - Lecturer edit - Started [is_empty]
 * [x] AUI Data - Lecturer edit - Autocomplete userID selected
 * [x] AUI Data - Lecturer edit - Autocomplete no results shown
 * [x] AUI Data - Lecturer edit - Freetext entered
 * [x] AUI Data - Lecturer edit - Item removed
 * [x] AUI Data - Lecturer edit - Completed [time_from_start]
 * [x] AUI Data - Location edit - Started
 * [x] AUI Data - Location edit - Completed
 * [x] AUI Data - Type changed [type]
 * [x] AUI Data - Event deleted
 * [x] AUI Data - Batch edit - Started
 * [x] AUI Data - Batch edit - Title edited
 * [x] AUI Data - Batch edit - TimeDate - Started
 * [x] AUI Data - Batch edit - TimeDate - Closed
 * [x] AUI Data - Batch edit - TimeDate - Week ticked on
 * [x] AUI Data - Batch edit - TimeDate - Week ticked off
 * [x] AUI Data - Batch edit - TimeDate - Day changed
 * [x] AUI Data - Batch edit - TimeDate - Start hour changed
 * [x] AUI Data - Batch edit - TimeDate - Start minute changed
 * [x] AUI Data - Batch edit - TimeDate - End hour changed
 * [x] AUI Data - Batch edit - TimeDate - End minute changed
 * [x] AUI Data - Batch edit - TimeDate - Added another day
 * [x] AUI Data - Batch edit - Lecturer edit - Started [is_empty]
 * [x] AUI Data - Batch edit - Lecturer edit - Autocomplete userID selected
 * [x] AUI Data - Batch edit - Lecturer edit - Autocomplete no results shown
 * [x] AUI Data - Batch edit - Lecturer edit - Freetext entered
 * [x] AUI Data - Batch edit - Lecturer edit - Item removed
 * [x] AUI Data - Batch edit - Lecturer edit - Completed [time_from_start]
 * [x] AUI Data - Batch edit - Location edited
 * [x] AUI Data - Batch edit - Type changed
 * [x] AUI Data - Batch edit - Cancelled [number_of_changes, number_of_selected_events, number_of_selected_terms, time_from_start, triposID]
 * [x] AUI Data - Batch edit - Completed [number_of_changes, number_of_selected_events, number_of_selected_terms, time_from_start, triposID]
 * [x] AUI Data - Batch edit - Clicked on disabled batch edit component

**Timers**

Critical user journeys and tasks should be timed via timestamps, and the delta between the start timestamp and the completed timestamp should be sent as metadata with the completed event.

*Student UI*
 * [x] Time to first module load (pageload > module list loaded)
 * [x] Time to first event in personal calendar (pageload > first event in calendar)
 * [x] Time to calendar export (pageload > Subscription or any export link clicked)
 
*Admin UI*
 * [x] Create module workflow
 * [x] Create series workflow
 * [x] Borrow series workflow
 * [x] Series edit workflow
 * [x] Batch edit sub-workflow
 * [x] Publish workflow

### CODE REVIEW

**Comment by bp323** [comment]
Converting the issue in a PR for transparency

**Comment by bp323** [comment]
Putting this on hold to verify with @oszkarnagy

File: apps/timetable/ui/index.html [hunk 1]
```diff
@@ -41,7 +41,7 @@
         <div id="gh-page-container">
             <div id="gh-left-container">
                 <div id="gh-meta-container">
-                    <a href="http://cam.ac.uk" id="gh-header-logo" target="_blank" title="Visit the Cambridge University website">
+                    <a href="javascript:;" id="gh-header-logo" target="_blank" title="Visit Timetable">
```
Comment (Coenego) [inline apps/timetable/ui/index.html]: These `target="_blank"` can probably go if we're not using the default behaviour anymore?

File: shared/gh/js/gh.bootstrap.js [hunk 2]
```diff
@@ -62,6 +62,7 @@ requirejs.config({
         'gh.constants': 'gh/js/constants/gh.constants',
 
         // GH utilities
+        'gh.utils.instrumentation': 'gh/js/utils/gh.utils.instrumentation',
```
Comment (Coenego) [inline shared/gh/js/gh.bootstrap.js]: This should go after `gh.utils`

File: shared/gh/js/utils/gh.utils.instrumentation.js [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an "AS IS"
+ * BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+define(['exports'], function(exports) {
+
+    var setUpInstrumentation = exports.setUpInstrumentation = function(me, callback) {
```
Comment (Coenego) [inline shared/gh/js/utils/gh.utils.instrumentation.js]: Some function documentation needed

File: shared/gh/js/views/gh.calendar.js [hunk 4]
```diff
@@ -50,6 +133,12 @@ define(['gh.core', 'moment', 'clickover'], function(gh, moment) {
         // Update the calendar
         calendar.fullCalendar(action);
 
+        if (action === 'next') {
+            sendNavigationEvent(true);
+        } else {
+            sendNavigationEvent();
+        }
+
```
Comment (Coenego) [inline shared/gh/js/views/gh.calendar.js]: This can probably be written shorter:

```
sendNavigationEvent(action === 'next')
```

**Comment by Coenego** [comment]
Nice work!! Only some minor comments remaining. Reassigning to @bp323

**Comment by bp323** [comment]
This depends on https://github.com/fronteerio/grasshopper/pull/82

**Comment by bp323** [comment]
The backend PR has been merged, this is ready for review.

File: shared/gh/js/utils/gh.utils.js [hunk 5]
```diff
@@ -305,10 +242,9 @@ define(['exports', 'gh.utils.state', 'gh.utils.templates', 'gh.utils.time', 'boo
             $('body').append($notificationContainer);
         }
 
-        // If a title has been provided, we wrap it in an h4 and prepend it to the message
-        if (title) {
-            message = '<div data-internal-id="' + randomId + '"><h4>' + title + '</h4>' + (message ? '<p>' + message + '</p>' : '') + '</div>';
-        }
+        // Wrap tit title in an h4 and prepend it to the message
+        /* istanbul ignore else */
```
Comment (Coenego) [inline shared/gh/js/utils/gh.utils.js]: I think the `/* istanbul ignore else */` can go?

File: shared/gh/js/views/gh.admin-batch-edit-organiser.js [hunk 6]
```diff
@@ -65,6 +69,29 @@ define(['gh.core', 'jquery-autosuggest'], function(gh) {
     };
 
     /**
+     * Track a selection event
+     *
+     * @param  {Object}    elem    The DOM element that was added after a selection
+     */
```
Comment (Coenego) [inline shared/gh/js/views/gh.admin-batch-edit-organiser.js]: `@private` missing

File: shared/gh/js/views/gh.admin-batch-edit-organiser.js [hunk 7]
```diff
@@ -208,10 +252,18 @@ define(['gh.core', 'jquery-autosuggest'], function(gh) {
      *
```
Comment (Coenego) [inline shared/gh/js/views/gh.admin-batch-edit-organiser.js]: `@param` missing

**Comment by Coenego** [comment]
Some minor issues left. Reassigning to @bp323

**Comment by bp323** [comment]
Followed up on review, back to @Coenego

**Pull request merged**